src/Nethermind/Nethermind.Trie/PatriciaTree.cs

```diff
@@ -1168,17 +1245,47 @@ public void Accept<TNodeContext>(ITreeVisitor<TNodeContext> visitor, Hash256 roo
             ITrieNodeResolver resolver = TrieStore;
             if (flags != ReadFlags.None)
             {
-                resolver = new TrieNodeResolverWithReadFlags(TrieStore, flags);
+                resolver = new TrieNodeResolverWithReadFlags(resolver, flags);
             }
 
-            visitor.VisitTree(default, rootHash, trieVisitContext);
+            if (!visitor.IsFullDbScan)
```

Comment:
this method from here is very hard to follow, multiple if's and returns, can you refactor for it to be more understandable?